File: algorithms/M/Magnetic Force Between Two Balls/Magnetic Force Between Two Balls.java
class Solution {
    
    boolean isPossible(int[] position, int m, int mid)
    {
        int count = 1;
        int placed = position[0];
        for(int i=1; i<position.length; i++)
        {
            if(position[i]-placed >= mid)
            {
                placed = position[i];
                count++;
            }
            if(count==m)
            {
                return true;
            }
        }
        return false;
    }
    
    public int maxDistance(int[] position, int m) {
        
        Arrays.sort(position);
        int n = position.length;
        
        int s = 1;
        int e = position[n-1];
        int ans = 0;
        
        while(s<=e)
        {
            int mid = (s+e)/2;
            if(isPossible(position, m, mid))
            {
                ans = mid;
                s = mid+1;
            }
            else
            {
                e = mid-1;
            }
        }
        
        return ans;
        
    }
}

File: algorithms/M/Magnetic Force Between Two Balls/Magnetic Force Between Two Balls.cpp
class Solution {
public:
    bool isPossible(vector<int>& position, int m,int mid){
        long long int basketCount=1;
        int lastPos=position[0];
        for(int i=0;i<position.size();i++){
            if((position[i]-lastPos)>=mid){
                basketCount++;
                lastPos=position[i];
            }
        }
        return basketCount>=m;
    }
    int maxDistance(vector<int>& position, int m) {
        sort(position.begin(),position.end());
        int n=position.size();
        long long int start=1;
        long long int end=position[n-1]-position[0];
        int ans=0;
        int mid= start+(end-start)/2;
        while(start<=end){
            if(isPossible(position,m,mid)){
            ans=mid;
            start=mid+1;
        }
        else
            end=mid-1;
           mid= start+(end-start)/2; 
        }
        return ans;
    }
};